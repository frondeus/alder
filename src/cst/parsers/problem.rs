use crate::*;

pub fn recover(parser: impl Parser) -> impl Parser {
    move |state: &mut State| {
        let node = parser.parse(state);
        if !node.is(NodeId::ERROR) {
            state.panic = false;
        }
        node
    }
}

pub fn raise(problem: impl Problem + Clone + 'static, len: usize) -> impl Parser {
    move |state: &mut State| {
        let panic = state.panic;
        let span = state.input.chomp(len);
        match state.last_error() {
            Some(err) if panic => {
                err.span.range.1 += len;
                if let Some(error) = state.errors.last_mut() {
                    error.span.range.1 += len;
                }
                none().parse(state)
            }
            _ if !panic => {
                let problem = Box::new(problem.clone()) as Box<dyn Problem + 'static>;
                let context = state
                    .nodes
                    .iter()
                    .flat_map(|node| node.all_names_with_span())
                    .filter(|(name, _)| !NodeId::NO_CONTEXT.contains(name))
                    .map(|(name, span)| ParseErrorContext::new(name, span))
                    .collect();
                state
                    .errors
                    .push(ParseError::new(problem, span.clone(), context));
                state.panic = true;
                Node::error(span)
            }
            _ => Node::error(span),
        }
    }
}
