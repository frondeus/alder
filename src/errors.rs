use log::error;
use std::error::Error as StdError;

struct SourceIter<'a> {
    current: Option<&'a (dyn StdError + 'static)>,
}
impl<'a> Iterator for SourceIter<'a> {
    type Item = &'a (dyn StdError + 'static);
    fn next(&mut self) -> Option<Self::Item> {
        let current = self.current;
        self.current = self.current.and_then(StdError::source);
        current
    }
}

pub fn source_iter(error: &impl StdError) -> impl Iterator<Item = &(dyn StdError + 'static)> {
    SourceIter {
        current: error.source(),
    }
}

pub trait ResultExt<T> {
    fn unwrap_display(self) -> T;
    fn unwrap_display_err(self) -> String;
    fn unwrap_log(self) -> Option<T>;
}

impl<T, E> ResultExt<T> for Result<T, E>
where
    E: StdError,
{
    fn unwrap_display(self) -> T {
        match self {
            Err(err) => {
                let mut cause = format!("{}", err);
                for source in source_iter(&err) {
                    cause += &format!("\ncaused by: {}", source);
                }
                panic!("\n{}", cause)
            }
            Ok(o) => o,
        }
    }

    fn unwrap_display_err(self) -> String {
        if let Err(err) = self {
            let mut cause = format!("{}", err);
            for source in source_iter(&err) {
                cause += &format!("\ncaused by: {}", source);
            }
            cause
        } else {
            panic!("Expected error")
        }
    }

    fn unwrap_log(self) -> Option<T> {
        match self {
            Err(err) => {
                error!("{}", err);
                for source in source_iter(&err) {
                    error!("caused by: {}", source);
                }
                None
            }
            Ok(o) => Some(o),
        }
    }
}
