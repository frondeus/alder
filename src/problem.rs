use crate::offset::Offset;
use crate::{Parser, Result};
use colored::{ColoredString, Colorize};
use std::fmt::Error;
use std::fmt::Formatter;
use std::fmt::{Debug, Display};
use std::iter::repeat;
use std::ops::{Deref, DerefMut};

pub struct Ctx<P1, C> {
    pub(crate) c: C,
    pub(crate) p: P1,
}
impl<'a, P, C, P1> Parser<'a, C, P> for Ctx<P1, C>
where
    P1: Parser<'a, C, P>,
    C: Copy,
{
    type T = P1::T;

    fn parse(&self, i: &'a str) -> Result<'a, Self::T, C, P> {
        self.p.parse(i).map_err(|mut e| {
            e.iter_mut().for_each(|e| e.add_context(self.c, i));
            e
        })
    }
}

pub fn context<'a, P1, P, C, T>(c: C, p: P1) -> impl Parser<'a, C, P, T = T>
where
    P1: Parser<'a, C, P, T = T>,
    C: Copy,
{
    Ctx { c, p }
}

#[derive(Debug)]
pub struct DeadEnds<'a, C, P> {
    pub ends: Vec<DeadEnd<'a, C, P>>,
}

impl<'a, C, P> DeadEnds<'a, C, P> {
    pub fn failure(&self) -> bool {
        self.ends.iter().any(|e| e.failure)
    }
}

impl<'a, P, C> Deref for DeadEnds<'a, C, P> {
    type Target = Vec<DeadEnd<'a, C, P>>;

    fn deref(&self) -> &Self::Target {
        &self.ends
    }
}

impl<'a, P, C> DerefMut for DeadEnds<'a, C, P> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.ends
    }
}

impl<'a, P, C> From<Vec<DeadEnd<'a, C, P>>> for DeadEnds<'a, C, P> {
    fn from(ends: Vec<DeadEnd<'a, C, P>>) -> Self {
        Self { ends }
    }
}

#[derive(Debug)]
pub struct DeadEnd<'a, C, P> {
    pub problem: P,
    pub failure: bool,
    pub input: &'a str,
    pub context: Vec<(&'a str, C)>,
}

impl<'a, P, C> DeadEnd<'a, C, P> {
    pub fn new(problem: P, input: &'a str) -> Self {
        Self {
            problem,
            input,
            failure: false,
            context: vec![],
        }
    }

    pub fn vec(problem: P, input: &'a str) -> DeadEnds<'a, C, P> {
        DeadEnds {
            ends: vec![Self::new(problem, input)],
        }
    }

    pub fn add_context(&mut self, c: C, input: &'a str) {
        //if !self.failure {
        //self.context.insert(0, (input, c));
        self.context.push((input, c));
        //}
    }

    pub fn cut(&mut self) {
        self.failure = true;
    }
}

pub fn cut<'a, C, P, T>(p: impl Parser<'a, C, P, T = T>) -> impl Parser<'a, C, P, T = T> {
    move |i| {
        p.parse(i).map_err(|mut d| {
            d.iter_mut().for_each(|i| i.cut());
            d
        })
    }
}

pub struct DisplayInputEntry<'a> {
    whre: &'a str,
    desc: Option<ColoredString>,
}

pub struct DisplayInput<'a> {
    input: &'a str,
    entries: Vec<DisplayInputEntry<'a>>,
}

impl<'a> DisplayInput<'a> {
    pub fn new(input: &'a str, whre: &'a str) -> Self {
        Self {
            input,
            entries: vec![DisplayInputEntry { whre, desc: None }],
        }
    }
    pub fn with_desc(mut self, desc: impl Into<ColoredString>) -> Self {
        self.entries[0].desc = Some(desc.into()); //TODO
        self
    }

    fn eol(line: usize, lines: &[&'a str]) -> &'static str {
        if lines.len() <= line + 1 {
            "\\EOF"
        } else {
            "\\n"
        }
    }

    /// Output: (line, column)
    fn position(&self, lines: &[&'a str], whre: &'a str) -> (usize, usize) {
        let mut offset = self.input.offset(whre);
        let mut line = 0;
        let mut column = 0;

        for (j, l) in lines.iter().enumerate() {
            if offset <= l.len() {
                line = j;
                column = offset;
                break;
            } else {
                offset = offset - l.len() - 1;
                line = j + 1;
            }
        }

        (line, column)
    }
}

impl<'a> Display for DisplayInput<'a> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::result::Result<(), Error> {
        let input = self.input;
        let lines = input.lines().collect::<Vec<_>>();

        for entry in self.entries.iter() {
            let (line, column) = self.position(&lines, entry.whre);

            if line > 0 {
                if let Some(prev) = lines.get(line - 1) {
                    write!(f, "{}", format!("{}  | ", line - 1).cyan())?;
                    write!(f, "{}", (prev).bright_white())?;
                    writeln!(f, "{}", "\\n".bright_black())?;
                }
            }

            write!(f, "{}", format!("{}  | ", line).cyan())?;
            if let Some(this) = lines.get(line) {
                write!(f, "{}", this.bright_white())?;
            }
            writeln!(f, "{}", Self::eol(line, &lines).bright_black())?;
            write!(f, "{}", "   | ".cyan())?;
            if column > 0 {
                write!(f, "{}", &repeat(' ').take(column).collect::<String>())?;
            }
            let desc = entry.desc.clone().unwrap_or_else(|| "^".normal());
            writeln!(f, "{}", &desc.bold())?;
            if let Some(next) = lines.get(line + 1) {
                write!(f, "{}", format!("{}  | ", line + 1).cyan())?;
                write!(f, "{}", (next).bright_white())?;
                writeln!(f, "{}", Self::eol(line + 1, &lines).bright_black())?;
            }
        }

        Ok(())
    }
}
