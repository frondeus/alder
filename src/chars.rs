use crate::problem::DeadEnd;
use crate::Parser;

pub fn take<'a, C, P: Clone>(len: usize, problem: P) -> impl Parser<'a, C, P, T = &'a str> {
    move |i: &'a str| {
        let i_len = i.len();
        if i_len < len {
            return Err(DeadEnd::vec(problem.clone(), i));
        }
        Ok((&i[0..len], &i[len..]))
    }
}

pub fn char_where<'a, C, P: Clone>(
    f: impl Fn(char) -> bool,
    problem: P,
) -> impl Parser<'a, C, P, T = char> {
    let parser = take(1, problem.clone()).map(|s| s.chars().next());
    move |i| {
        let (first, rest) = parser.parse(i)?;
        match first {
            Some(letter) if f(letter) => Ok((letter, rest)),
            _ => Err(DeadEnd::vec(problem.clone(), i)),
        }
    }
}

pub fn chomp_while<'a, C, P>(f: impl Fn(char) -> bool) -> impl Parser<'a, C, P, T = &'a str> {
    move |i: &'a str| {
        let mut len = 0;
        loop {
            let c = &i[len..].chars().next();
            match c {
                Some(letter) if f(*letter) => {
                    len += 1;
                }
                _ => break,
            }
        }

        let result = &i[0..len];
        let rest = &i[len..];
        Ok((result, rest))
    }
}

pub fn token<'a, C, P: Clone>(expected: char, problem: P) -> impl Parser<'a, C, P, T = char> {
    let parser = take(1, problem.clone()).map(|s| s.chars().next());
    move |i| {
        let (first, rest) = parser.parse(i)?;
        match first {
            Some(letter) if letter == expected => Ok((letter, rest)),
            _ => Err(DeadEnd::vec(problem.clone(), i)),
        }
    }
}

pub fn tag<C, P: Clone>(expected: &str, problem: P) -> impl Parser<C, P, T = &str> {
    move |i| {
        let len = expected.len();
        let (o, r) = take(len, problem.clone()).parse(i)?;
        if o == expected {
            return Ok((o, r));
        }
        Err(DeadEnd::vec(problem.clone(), i))
    }
}

pub fn spaces<'a, C, P>() -> impl Parser<'a, C, P, T = &'a str> {
    chomp_while(|f| f.is_whitespace())
}
/*
impl<'a, C> Parser<'a, C, CharProblem<'a>> for char {
    type T = char;

    fn parse(&self, i: &'a str) -> Result<'a, Self::T, C, CharProblem<'a>> {
        token(*self).parse(i)
    }
}


impl<'a, C> Parser<'a, C, CharProblem<'a>> for &'a str {
    type T = &'a str;

    fn parse(&self, i: &'a str) -> Result<'a, Self::T, C, CharProblem<'a>> {
        tag(self).parse(i)
    }
}
*/
