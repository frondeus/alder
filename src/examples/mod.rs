use super::*;
use crate::chars::{spaces, token};
use crate::errors::ResultExt;
use crate::examples::report::Report;
use crate::problem::cut;
use colored::Colorize;
use derive_more::Display;
use std::fmt::Debug;
use test_case::test_case;

#[derive(Debug)]
enum Cons {
    Cons,
    Nil,
    Pair(Box<Cons>, Box<Cons>),
}

#[derive(Debug, Display, Clone, PartialEq)]
pub enum Problem {
    #[display(fmt = "I expected 'nil'")]
    ExpectedNil,

    #[display(fmt = "I expected '('")]
    ExpectedOpenBracket,

    #[display(fmt = "I expected '.'")]
    ExpectedDot,

    #[display(fmt = "I expected ')'")]
    ExpectedCloseBracket,
}

impl Problem {
    pub fn name(&self) -> &'static str {
        use Problem::*;
        match self {
            ExpectedNil => "()",
            ExpectedOpenBracket => "(",
            ExpectedCloseBracket => ")",
            ExpectedDot => ".",
        }
    }
}

#[derive(Debug, Clone, Copy, Display)]
pub enum ConsContext {
    #[display(fmt = "I tried to parse nil value")]
    Nil,
    #[display(
        fmt = "I tried to parse cons pair {}",
        r#""(<expr> . <expr>)".white().bold()"#
    )]
    Pair,
    #[display(fmt = "I tried to parse expression")]
    Expression,
}

mod report;

fn nil<'a>() -> impl Parser<'a, ConsContext, Problem, T = Cons> {
    //tag("()", Problem::ExpectedNil)
    (
        token('(', Problem::ExpectedOpenBracket),
        token(')', Problem::ExpectedCloseBracket),
    )
        .ctx(ConsContext::Nil)
        .map(|_| Cons::Nil)
}

fn cons<'a>() -> impl Parser<'a, ConsContext, Problem, T = Cons> {
    |i| preceded(spaces(), or(nil(), pair()).ctx(ConsContext::Expression)).parse(i)
}

fn pair<'a>() -> impl Parser<'a, ConsContext, Problem, T = Cons> {
    (
        token('(', Problem::ExpectedOpenBracket),
        cons(),
        spaces(),
        token('.', Problem::ExpectedDot),
        cut((cons(), spaces(), token(')', Problem::ExpectedCloseBracket))),
    )
        .map(|(_, lhs, _, _, (rhs, _, _))| Cons::Pair(Box::new(lhs), Box::new(rhs)))
        .ctx(ConsContext::Pair)
}

#[test_case(cons(),   "("    ; "eof" )]
#[test_case(cons(),   r#"(
    ()
    .
    (a
)"#   ; "pair error" )]
#[test_case(cons(),   r#"(
    ()
    .
    ()
"#  ; "missing close paren" )]
fn test_errors<'a, T: Debug>(p: impl Parser<'a, ConsContext, Problem, T = T>, input: &'a str) {
    let err = p
        .parse(input)
        .map_err(|ends| Report::new(input, ends))
        .unwrap_display_err();

    insta::assert_display_snapshot!(err);
}

#[test_case(cons(),   "()"   => matches ( Cons::Nil, _) )]
fn test<'a, T: Debug>(
    p: impl Parser<'a, ConsContext, Problem, T = T>,
    input: &'a str,
) -> (T, &'a str) {
    p.parse(input)
        .map_err(|ends| Report::new(input, ends))
        .unwrap_display()
}
