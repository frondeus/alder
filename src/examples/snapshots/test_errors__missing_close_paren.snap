---
source: src/examples/mod.rs
expression: err
---
[1;31m-- SYNTAX ERROR --[0m
I tried to parse cons pair [1;37m(<expr> . <expr>)[0m
[36m0  | [0m[97m([0m[90m\n[0m
[36m   | [0m[1;36m^ started here[0m
[36m1  | [0m[97m    ()[0m[90m\n[0m
[36m3  | [0m[97m    ()[0m[90m\n[0m
[36m4  | [0m[90m\EOF[0m
[36m   | [0m[1m[31m^ unexpected token.[0m[1m I expected ')'[0m


